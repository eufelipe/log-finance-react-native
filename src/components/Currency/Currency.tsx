import React, {useCallback, useState} from 'react';
import NumberFormat from 'react-number-format';
import {Text} from 'styles/layout';
import CurrencyService, {DEFAULT_CURRENCY} from 'services/currency';
import {useTranslation} from 'react-i18next';
import {DEFAULT_LANGUAGE} from 'services/language';
import {useFocusEffect} from '@react-navigation/native';

interface Props {
  value: number;
  decimalScale?: number;
  render?: (text: string) => React.ReactNode | string;
}

export function currencyFormatter(
  value: number,
  currency?: string,
  language?: string,
): string {
  let output = 0;
  if (Number(value)) {
    output = value;
  }

  const currentCurrency = currency ?? DEFAULT_CURRENCY;
  const currentLanguage = language ?? DEFAULT_LANGUAGE;

  const optionsNumber = {
    style: 'currency',
    currency: currentCurrency,
  };

  const amount = new Intl.NumberFormat(currentLanguage, optionsNumber).format(
    output / 100,
  );

  return `${amount}`;
}

const Currency = ({value, decimalScale = 2, render}: Props): JSX.Element => {
  const [currentCurrency, setCurrentCurrency] = useState<string>();
  const {i18n} = useTranslation();

  const loadCurrentCurrency = useCallback(async () => {
    const currentStored = await CurrencyService.getCurrency();
    if (currentStored) setCurrentCurrency(currentStored);
  }, []);

  const renderItem = (text: string) => {
    if (render) {
      return render(text);
    }
    return <Text>{text}</Text>;
  };

  const format = currencyFormatter(value, currentCurrency, i18n.language);

  useFocusEffect(
    useCallback(() => {
      loadCurrentCurrency();
    }, [loadCurrentCurrency]),
  );

  return (
    <NumberFormat
      fixedDecimalScale
      format={format}
      displayType="text"
      value={value}
      decimalScale={decimalScale}
      renderText={renderItem}
    />
  );
};

export default Currency;
