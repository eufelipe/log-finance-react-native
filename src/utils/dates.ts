import {
  format as formatPattern,
  isValid,
  parseISO as parseISOFNS,
} from 'date-fns';

import locale from 'date-fns/locale/pt-BR';

export const DATE_FORMAT_ISO = 'yyyy-MM-dd';
export const DATE_FORMAT_BR = 'dd/MM/yyyy';
export const DATE_FORMAT_FRIENDLY = " dd/MM 'às' HH:mm";
export const DATE_FORMAT_FRIENDLY_TIME = "'às' HH:mm";

export function parseStringToDate(value: string): Date {
  const parsedDate = parseISOFNS(value);
  if (!isValid(parsedDate)) {
    throw Error('canot parse date');
  }
  return parsedDate;
}
export function parseDateToString({
  value,
  format = DATE_FORMAT_BR,
}: {
  value: Date;
  format?: string;
}): string {
  return formatPattern(value, format, {locale});
}

export const getDate = (date: Date): string => {
  return formatPattern(date, DATE_FORMAT_ISO, {locale});
};

export const getDateToday = (): string => {
  const today = new Date();
  return formatPattern(today, DATE_FORMAT_ISO, {locale});
};

export const getDateAndHourFriendly = (value: Date): string => {
  return formatPattern(value, DATE_FORMAT_FRIENDLY, {locale});
};

export const getTimeFriendly = (value: Date): string => {
  return formatPattern(value, DATE_FORMAT_FRIENDLY_TIME, {locale});
};
