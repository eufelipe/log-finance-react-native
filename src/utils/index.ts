export {default as number} from './number';
export {default as strings} from './strings';
