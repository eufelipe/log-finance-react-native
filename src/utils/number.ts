export const sanitizeToFloat = (value: string): number => {
  value = value.replace(/[^0-9.]+/, '');
  if (value.includes('.') && value.includes(',')) {
    value = value.replace(/\./g, '').replace(',', '.');
  } else if (value.includes(',')) {
    value = value.replace(',', '.');
  }
  return parseFloat(value) || 0;
};
