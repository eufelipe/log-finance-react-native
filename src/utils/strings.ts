export const sanitizeString = (value: string): string => {
  return value
    .toLowerCase()
    .replace(/á/g, 'a')
    .replace(/é/g, 'e')
    .replace(/í/g, 'i')
    .replace(/ó/g, 'o')
    .replace(/ú/g, 'u');
};
