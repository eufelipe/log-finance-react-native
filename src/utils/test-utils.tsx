/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import {render, RenderAPI, RenderOptions} from '@testing-library/react-native';

import {theme} from 'styles/theme';
import {ThemeProvider} from 'styled-components/native';

import {NavigationContainer} from '@react-navigation/native';

interface Props {
  children: JSX.Element;
}

const AllTheProviders = ({children}: Props) => {
  return (
    <ThemeProvider theme={theme}>
      <NavigationContainer>{children}</NavigationContainer>
    </ThemeProvider>
  );
};

const customRender = (
  ui: React.ReactElement<any>,
  options?: RenderOptions,
): RenderAPI => render(ui, {wrapper: AllTheProviders, ...options});

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
