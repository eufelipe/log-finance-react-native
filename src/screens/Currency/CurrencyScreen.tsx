import React, {useCallback, useMemo, useEffect, useState} from 'react';
import {useNavigation} from '@react-navigation/native';

import {Container} from './styles';
import {useTranslation} from 'react-i18next';

import List, {ListItemProps} from 'components/List';
import {Toolbar} from 'components';

import {filter, sortBy} from 'lodash';

import currencies, {CurrencyItem} from './currencies';
import CurrencyService from 'services/currency';
import {sanitizeString} from 'utils/strings';

const CurrencyScreen = (): JSX.Element => {
  const navigation = useNavigation();
  const [query, setQuery] = useState('');
  const [currentCurrency, setCurrentCurrency] = useState<string>();
  const [filteredCurrencies, setFilteredCurrencies] = useState<ListItemProps[]>(
    [],
  );

  const {t} = useTranslation('currencies');

  const changeCurrency = useCallback(
    async (currency: CurrencyItem) => {
      await CurrencyService.saveCurrency(currency.code);

      navigation.goBack();
    },
    [navigation],
  );

  const loadCurrentCurrency = useCallback(async () => {
    const currentStored = await CurrencyService.getCurrency();
    if (currentStored) setCurrentCurrency(currentStored);
  }, []);

  const currencyItems: ListItemProps[] = useMemo(() => {
    const items = currencies.map((currency, index) => ({
      id: index + 1,
      title: currency.name,
      subtitle: currency.code,
      selected: currentCurrency === currency.code ? t('active') : undefined,
      icon: 'chevron-right',
      onPress: () => changeCurrency(currency),
    }));

    const ordered = sortBy(items, 'selected');
    return ordered;
  }, [changeCurrency, currentCurrency, t]);

  const handleSearch = (term: string) => {
    const formattedQuery = sanitizeString(term);
    const filtered = filter(currencyItems, ({title, subtitle}) => {
      return (
        sanitizeString(title).includes(formattedQuery) ||
        (!!subtitle && sanitizeString(subtitle).includes(formattedQuery))
      );
    });
    setFilteredCurrencies(filtered);
    setQuery(term);
  };

  useEffect(() => {
    loadCurrentCurrency();
  }, [loadCurrentCurrency]);

  useEffect(() => {
    if (!query) {
      setFilteredCurrencies(currencyItems);
    }
  }, [query, currencyItems]);

  return (
    <Container>
      <Toolbar title={t('title')} query={query} onChangeText={handleSearch} />
      <List items={filteredCurrencies} />
    </Container>
  );
};

export default CurrencyScreen;
