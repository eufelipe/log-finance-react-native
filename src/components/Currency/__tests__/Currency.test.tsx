/* eslint-disable @typescript-eslint/unbound-method */
import React from 'react';
import {Text} from 'react-native';
import {render} from 'utils/test-utils';

import Currency from '../Currency';

describe('Currency', () => {
  it('Test match snapshot Currency', () => {
    // given
    const props = {
      value: 100.28,
      render: (text: string) => <Text>{text}</Text>,
    };

    // when
    const {toJSON} = render(<Currency {...props} />);

    // then
    expect(toJSON()).toMatchSnapshot();
  });

  it('Test if the value renders correctly', () => {
    // given
    const props = {
      value: 99.45,
      render: (text: string) => <Text>{text}</Text>,
    };

    // when
    const {getByText} = render(<Currency {...props} />);

    // then
    expect(getByText('R$ 99,45')).toBeTruthy();
    expect(() => getByText(/20,45/i)).toThrow('Unable to find an element');
  });
});
