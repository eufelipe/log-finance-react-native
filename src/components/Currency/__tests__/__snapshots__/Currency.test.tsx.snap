// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency Test match snapshot Currency 1`] = `
<Text>
  R$ 100,28
</Text>
`;
